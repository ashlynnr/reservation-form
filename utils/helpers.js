export const orderHandler = ({ id, value, key, reviseOrder, order }) => {
  const updatedOrder = order.map(a => Object.assign({}, a));

  updatedOrder.map(data => {
    if (data.id === id) {
      data[key] = value;
    }
  });
  reviseOrder(updatedOrder);
  return updatedOrder;
};

export const removeRooms = ({ reviseOrder, order, room, index }) => {
  const updatedOrder = order.map(a => Object.assign({}, a));
  let removals = 1;
  updatedOrder.map(data => {
    if (data.id > room.id) {
      removals++;
    }
  });
  updatedOrder.splice(index, index + removals);
  reviseOrder(updatedOrder);
  return updatedOrder;
};

export const addRooms = ({ reviseOrder, order, room }) => {
  const updatedOrder = order.map(a => Object.assign({}, a));
  let firstRoom = 1;
  const lastRoom = room.id;
  let additions = [];

  while (firstRoom <= lastRoom) {
    additions.push(firstRoom++);
  }

  additions.map(roomId => {
    if (!order.some(data => data.id === roomId)) {
      updatedOrder.push({ id: roomId, adults: 1, children: 0 });
    }
  });
  reviseOrder(updatedOrder);
  return updatedOrder;
};
