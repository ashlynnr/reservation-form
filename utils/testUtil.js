export const mockRooms = [
  {
    id: 1,
    name: "Room 1",
    adults: [1, 2],
    children: [0, 1, 2],
    optional: false
  },
  {
    id: 2,
    name: "Room 2",
    adults: [1, 2],
    children: [0, 1, 2],
    optional: true
  },
  {
    id: 3,
    name: "Room 3",
    adults: [1, 2],
    children: [0, 1, 2],
    optional: true
  },
  {
    id: 4,
    name: "Room 4",
    adults: [1, 2],
    children: [0, 1, 2],
    optional: true
  }
];
