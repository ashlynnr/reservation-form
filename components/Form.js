import React, { useEffect } from "react";
import styled from "styled-components";
import DropdownOptions from "./DropdownOptions";
import { editReservation } from "../utils/helpers";

const Form = ({ room, order, reviseOrder, index }) => {
  const checkHandler = e => {
    if (!e.target.checked) {
      removeRooms();
    } else {
      addRooms();
    }
  };

  const orderHandler = (id, value, key) => {
    const updatedOrder = order.map(a => Object.assign({}, a));

    updatedOrder.map((data, i) => {
      if (data.id === id) {
        data[key] = value;
      }
    });
    reviseOrder(updatedOrder);
  };

  const removeRooms = () => {
    const updatedOrder = order.map(a => Object.assign({}, a));
    let removals = 1;
    updatedOrder.map(data => {
      if (data.id > room.id) {
        removals++;
      }
    });
    updatedOrder.splice(index, index + removals);
    reviseOrder(updatedOrder);
  };

  const addRooms = () => {
    const updatedOrder = order.map(a => Object.assign({}, a));
    let firstRoom = 1;
    const lastRoom = room.id;
    let additions = [];

    while (firstRoom <= lastRoom) {
      additions.push(firstRoom++);
    }

    additions.map(roomId => {
      if (!order.some(data => data.id === roomId)) {
        updatedOrder.push({ id: roomId, adults: 1, children: 0 });
      }
    });
    reviseOrder(updatedOrder);
  };

  const checked = order.some(data => data.id === room.id);

  return (
    <Container checked={checked}>
      <Header>
        {room.optional && (
          <Checkbox
            type="checkbox"
            name="room"
            value={room.name}
            onClick={checkHandler}
            checked={checked}
          />
        )}
        {room.name}
      </Header>
      <Section checked={checked}>
        <DropdownOptions
          adults={room.adults}
          children={room.children}
          checked={checked}
          room={room}
          orderHandler={orderHandler}
          order={order}
        />
      </Section>
    </Container>
  );
};

export default Form;

const Container = styled.div`
  display: grid;
  grid-template-rows: 25% 75%;
  height: 100%;
  width: 100%;
  background: ${props =>
    props.checked ? props.theme.grey : props.theme.disabledLightPurple};
  border: ${props =>
    !props.checked
      ? `solid ${props.theme.disabledPurple} 5px`
      : `solid ${props.theme.grey} 5px`};
  border-radius: 2%;
  transition: 0.3s all ease-in;
`;

const Section = styled.div`
  height: 98%;
  background: ${props => (props.checked ? "white" : "transparent")};
  border-radius: 2%;
  grid-template-columns: 2fr;
  transition: 0.3s background ease-in;
`;

const Header = styled.div`
  font-weight: bold;
  font-size: 1.2em;
  height: 100%;
  display: flex;
  align-items: center;
  padding: 0 5px;
`;

const Checkbox = styled.input`
  -ms-transform: scale(1.5);
  -moz-transform: scale(1.5);
  -webkit-transform: scale(1.5);
  -o-transform: scale(1.5);
  transform: scale(1.5);
  margin-right: 10px;
`;
