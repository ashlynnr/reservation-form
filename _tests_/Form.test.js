import React from "react";
import { shallow } from "enzyme";
import Form from "../components/Form";
import { addRooms, removeRooms } from "../utils/helpers";

describe("Form", () => {
  // matches snapshot

  const defaultOrder = [{ id: 1, adults: 1, children: 0 }];
  const wrapper = shallow(
    <Form
      room={{ id: 2, adults: 1, children: 0 }}
      order={[{ id: 1, adults: 1, children: 0 }]}
      reviseOrder={jest.fn}
      index={1}
    />
  );

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("add Rooms runs correctly", () => {
    expect(
      addRooms({
        reviseOrder: jest.fn,
        order: [{ id: 1, adults: 2, children: 0 }],
        room: { id: 3, adults: 1, children: 0 }
      })
    ).toStrictEqual([
      { id: 1, adults: 2, children: 0 },
      { id: 2, adults: 1, children: 0 },
      { id: 3, adults: 1, children: 0 }
    ]);
  });

  it("remove Rooms to runs correctly", () => {
    expect(
      removeRooms({
        reviseOrder: jest.fn,
        order: [
          { id: 1, adults: 2, children: 0 },
          { id: 2, adults: 1, children: 0 },
          { id: 3, adults: 1, children: 0 }
        ],
        room: { id: 3, adults: 1, children: 0 },
        index: 1
      })
    ).toStrictEqual([{ id: 1, adults: 2, children: 0 }]);
  });
});
