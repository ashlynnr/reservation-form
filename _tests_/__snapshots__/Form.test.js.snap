// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `ShallowWrapper {}`;
